From 76ed3070ffa1490c5cb1b05d43a084758facc666 Mon Sep 17 00:00:00 2001
From: congqixia <congqi.xia@zilliz.com>
Date: Wed, 9 Aug 2023 11:19:15 +0800
Subject: [PATCH 2/4] Bump semver to 2.2.13 (#26206)

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>
---
 internal/common/version.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/internal/common/version.go b/internal/common/version.go
index 2b2ae8cd7..ee4affbf9 100644
--- a/internal/common/version.go
+++ b/internal/common/version.go
@@ -6,5 +6,5 @@ import "github.com/blang/semver/v4"
 var Version semver.Version
 
 func init() {
-	Version, _ = semver.Parse("2.2.12")
+	Version, _ = semver.Parse("2.2.13")
 }
-- 
2.41.0

