From 5dee8df08eadaeafaac2ad48ee0089cac1e7d4b4 Mon Sep 17 00:00:00 2001
From: Ji Bin <matrixji@live.com>
Date: Fri, 16 Jun 2023 11:06:10 +0800
Subject: [PATCH 4/4] patch for v2.2.10

Signed-off-by: Ji Bin <matrixji@live.com>
---
 internal/core/src/storage/MinioChunkManager.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/internal/core/src/storage/MinioChunkManager.cpp b/internal/core/src/storage/MinioChunkManager.cpp
index d60eb038c..f481c5a11 100644
--- a/internal/core/src/storage/MinioChunkManager.cpp
+++ b/internal/core/src/storage/MinioChunkManager.cpp
@@ -56,6 +56,7 @@ namespace milvus::storage {
 std::atomic<size_t> MinioChunkManager::init_count_(0);
 std::mutex MinioChunkManager::client_mutex_;
 
+#ifndef WIN32
 static void
 SwallowHandler(int signal) {
 #pragma GCC diagnostic push
@@ -71,6 +72,7 @@ SwallowHandler(int signal) {
     }
 #pragma GCC diagnostic pop
 }
+#endif
 
 /**
  * @brief convert std::string to Aws::String
@@ -105,6 +107,7 @@ MinioChunkManager::InitSDKAPI(RemoteStorageType type, bool useIAM, const std::st
     std::scoped_lock lock{client_mutex_};
     const size_t initCount = init_count_++;
     if (initCount == 0) {
+#ifndef WIN32
         // sdk_options_.httpOptions.installSigPipeHandler = true;
         struct sigaction psa;
         memset(&psa, 0, sizeof psa);
@@ -114,6 +117,7 @@ MinioChunkManager::InitSDKAPI(RemoteStorageType type, bool useIAM, const std::st
         sigemptyset(&psa.sa_mask);
         sigaddset(&psa.sa_mask, SIGPIPE);
         sigaction(SIGPIPE, &psa, 0);
+#endif
 #ifdef BUILD_GCP
         if (type == RemoteStorageType::GOOGLE_CLOUD && useIAM) {
             sdk_options_.httpOptions.httpClientFactory_create_fn = []() {
-- 
2.42.0

