diff --git a/internal/core/CMakeLists.txt b/internal/core/CMakeLists.txt
index eb1f727d2..bc47fe8a3 100644
--- a/internal/core/CMakeLists.txt
+++ b/internal/core/CMakeLists.txt
@@ -19,8 +19,6 @@ cmake_minimum_required( VERSION 3.18 )
 if ( APPLE )
     set( CMAKE_CROSSCOMPILING TRUE )
     set( RUN_HAVE_GNU_POSIX_REGEX 0 )
-    set( CMAKE_C_COMPILER "/usr/local/opt/llvm/bin/clang" )
-    set( CMAKE_CXX_COMPILER "/usr/local/opt/llvm/bin/clang++" )
 endif ()
 
 add_definitions(-DELPP_THREAD_SAFE)
@@ -31,7 +29,6 @@ project(core)
 include(CheckCXXCompilerFlag)
 if ( APPLE )
     message(STATUS "==============Darwin Environment==============")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I/usr/local/opt/llvm/include -I/usr/local/include -I/usr/local/opt/libomp/include -L/usr/local/opt/libomp/lib")
 elseif (${CMAKE_SYSTEM_NAME} MATCHES "Linux")
     message(STATUS "==============Linux Environment===============")
     set(LINUX TRUE)
diff --git a/scripts/core_build.sh b/scripts/core_build.sh
index 0c33d9439..3df09304e 100755
--- a/scripts/core_build.sh
+++ b/scripts/core_build.sh
@@ -187,12 +187,12 @@ fi
 unameOut="$(uname -s)"
 case "${unameOut}" in
     Darwin*)
-        llvm_prefix="$(brew --prefix llvm)"
+        llvm_prefix="$(brew --prefix llvm@14)"
         export CLANG_TOOLS_PATH="${llvm_prefix}/bin"
         export CC="${llvm_prefix}/bin/clang"
         export CXX="${llvm_prefix}/bin/clang++"
-        export LDFLAGS="-L${llvm_prefix}/lib -L/usr/local/opt/libomp/lib"
-        export CXXFLAGS="-I${llvm_prefix}/include -I/usr/local/include -I/usr/local/opt/libomp/include"
+        export LDFLAGS="-L${llvm_prefix}/lib -L$(brew --prefix libomp)/lib -L$(brew --prefix boost)/lib -L$(brew --prefix tbb)/lib"
+        export CXXFLAGS="-I${llvm_prefix}/include -I/usr/local/include -I$(brew --prefix libomp)/include -I$(brew --prefix boost)/include -I$(brew --prefix tbb)/include"
         ;;
           *)   echo "==System:${unameOut}";
 esac
diff --git a/scripts/install_deps.sh b/scripts/install_deps.sh
index 5d58bb3f5..7d66cd5ff 100755
--- a/scripts/install_deps.sh
+++ b/scripts/install_deps.sh
@@ -57,24 +57,15 @@ function install_linux_deps() {
 
 function install_mac_deps() {
   sudo xcode-select --install  > /dev/null 2>&1
-  brew install boost libomp ninja tbb cmake llvm ccache zstd
-  brew uninstall grep
+  brew install boost libomp ninja tbb cmake llvm@14 ccache zstd
   brew install grep
-  export PATH="/usr/local/opt/grep/libexec/gnubin:$PATH"
-  brew update && brew upgrade && brew cleanup
+  # export PATH="/usr/local/opt/grep/libexec/gnubin:$PATH"
+  # brew update && brew upgrade && brew cleanup
 
   if [[ $(arch) == 'arm64' ]]; then
     brew install openssl
     brew install librdkafka
     brew install pkg-config
-    sudo mkdir /usr/local/include
-    sudo mkdir /usr/local/opt
-    sudo ln -s "$(brew --prefix llvm)" "/usr/local/opt/llvm"
-    sudo ln -s "$(brew --prefix libomp)/include/omp.h" "/usr/local/include/omp.h"
-    sudo ln -s "$(brew --prefix libomp)" "/usr/local/opt/libomp"
-    sudo ln -s "$(brew --prefix boost)/include/boost" "/usr/local/include/boost"
-    sudo ln -s "$(brew --prefix tbb)/include/tbb" "/usr/local/include/tbb"
-    sudo ln -s "$(brew --prefix tbb)/include/oneapi" "/usr/local/include/oneapi"
   fi
 }
 
